File: harmonyos/src/selector.ts
// selector.ts — ChatCapability flags + DeviceAwareModelSelector.

import { DeviceSnapshot, DeviceTier, IDeviceContext } from './device';

/** Bitwise-composable capability flags. */
export const ChatCapability = {
  None:      0,
  Text:      1 << 0,
  Tools:     1 << 1,
  Vision:    1 << 2,
  Audio:     1 << 3,
  LongCtx:   1 << 4,
  Reasoning: 1 << 5,
  Streaming: 1 << 6,
  Default:   (1 << 0) | (1 << 6),  // Text | Streaming
} as const;

export function parseCapabilities(raw: string | null | undefined): number {
  if (!raw) return 0;
  let out = 0;
  for (const tok of raw.split(/[,\s]+/)) {
    switch (tok.trim().toLowerCase()) {
      case '': break;
      case 'text': out |= ChatCapability.Text; break;
      case 'tools': out |= ChatCapability.Tools; break;
      case 'vision': out |= ChatCapability.Vision; break;
      case 'audio': out |= ChatCapability.Audio; break;
      case 'longctx': case 'long_ctx': case 'long-ctx': out |= ChatCapability.LongCtx; break;
      case 'reasoning': out |= ChatCapability.Reasoning; break;
      case 'streaming': out |= ChatCapability.Streaming; break;
    }
  }
  return out;
}

export interface ModelCandidate {
  readonly name: string;
  readonly totalBytes: number;
  readonly capabilities: string | null;
}

export interface ModelSelection<T extends ModelCandidate = ModelCandidate> {
  readonly entry: T;
  readonly reason: string;
}

export interface IModelSelector<T extends ModelCandidate = ModelCandidate> {
  select(candidates: ReadonlyArray<T>, device: DeviceSnapshot, required: number): ModelSelection<T> | null;
}

function maxBytesForTier(tier: DeviceTier): number {
  switch (tier) {
    case DeviceTier.Wearable:    return 200 * 1024 * 1024;
    case DeviceTier.Embedded:    return 500 * 1024 * 1024;
    case DeviceTier.Phone:       return 2_500_000_000;
    case DeviceTier.Tablet:      return 6_000_000_000;
    case DeviceTier.Laptop:      return 20_000_000_000;
    case DeviceTier.Workstation: return 60_000_000_000;
  }
}

export class DeviceAwareModelSelector<T extends ModelCandidate = ModelCandidate> implements IModelSelector<T> {
  constructor(private readonly deviceContext: IDeviceContext) {}

  select(candidates: ReadonlyArray<T>, device: DeviceSnapshot, required: number = ChatCapability.Default): ModelSelection<T> | null {
    const ceil = maxBytesForTier(device.tier);
    const viable = candidates
      .filter(c => {
        const caps = parseCapabilities(c.capabilities ?? 'Text,Streaming');
        return (caps & required) === required && c.totalBytes <= ceil;
      })
      .sort((a, b) => a.totalBytes - b.totalBytes);
    if (viable.length === 0) return null;
    const chosen = viable[0];
    const reason = `tier=${DeviceTier[device.tier]} ram=${Math.floor(device.ramBytes / (1024 * 1024))}MB required=${required} → ${chosen.name} (${chosen.totalBytes} bytes)`;
    return { entry: chosen, reason };
  }
}

File: rust/src/selector.rs
//! selector.rs
//!
//! ChatCapability bitflags, IModelSelector, DeviceAwareModelSelector.

use bitflags::bitflags;
use serde::{Deserialize, Serialize};

use crate::catalog::ModelEntry;
use crate::device::{DeviceSnapshot, DeviceTier, IDeviceContext};

bitflags! {
    /// What features a chat model can do. Bitwise composable.
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
    pub struct ChatCapability: u32 {
        const NONE        = 0b0000_0000;
        const TEXT        = 0b0000_0001;
        const TOOLS       = 0b0000_0010;
        const VISION      = 0b0000_0100;
        const AUDIO       = 0b0000_1000;
        const LONG_CTX    = 0b0001_0000;
        const REASONING   = 0b0010_0000;
        const STREAMING   = 0b0100_0000;
        const DEFAULT     = Self::TEXT.bits() | Self::STREAMING.bits();
    }
}

/// Parse a "Capabilities" field from a catalog entry (comma- or space-separated).
pub fn parse_capabilities(raw: &str) -> ChatCapability {
    let mut out = ChatCapability::NONE;
    for tok in raw.split(|c: char| c == ',' || c.is_whitespace()) {
        match tok.trim().to_ascii_lowercase().as_str() {
            "" => continue,
            "text" => out |= ChatCapability::TEXT,
            "tools" => out |= ChatCapability::TOOLS,
            "vision" => out |= ChatCapability::VISION,
            "audio" => out |= ChatCapability::AUDIO,
            "longctx" | "long_ctx" | "long-ctx" => out |= ChatCapability::LONG_CTX,
            "reasoning" => out |= ChatCapability::REASONING,
            "streaming" => out |= ChatCapability::STREAMING,
            _ => {}
        }
    }
    out
}

/// Outcome of a selector call.
#[derive(Debug, Clone)]
pub struct ModelSelection {
    pub entry: ModelEntry,
    pub reason: String,
}

/// Strategy that picks a model given a device snapshot + required capability set.
pub trait IModelSelector: Send + Sync {
    fn select(
        &self,
        candidates: &[ModelEntry],
        device: &DeviceSnapshot,
        required: ChatCapability,
    ) -> Option<ModelSelection>;
}

/// Default selector. Filters by capability + tier ceiling, picks smallest
/// `total_bytes` that fits.
pub struct DeviceAwareModelSelector<'a> {
    pub device_context: &'a dyn IDeviceContext,
}

impl<'a> DeviceAwareModelSelector<'a> {
    pub fn new(device_context: &'a dyn IDeviceContext) -> Self {
        Self { device_context }
    }
}

impl<'a> IModelSelector for DeviceAwareModelSelector<'a> {
    fn select(
        &self,
        candidates: &[ModelEntry],
        device: &DeviceSnapshot,
        required: ChatCapability,
    ) -> Option<ModelSelection> {
        let max_bytes_for_tier = match device.tier {
            DeviceTier::Wearable => 200 * 1024 * 1024,
            DeviceTier::Embedded => 500 * 1024 * 1024,
            DeviceTier::Phone => 2_500_000_000,
            DeviceTier::Tablet => 6_000_000_000,
            DeviceTier::Laptop => 20_000_000_000,
            DeviceTier::Workstation => 60_000_000_000,
        };

        let mut viable: Vec<&ModelEntry> = candidates
            .iter()
            .filter(|c| {
                let caps = parse_capabilities(c.capabilities.as_deref().unwrap_or("Text,Streaming"));
                caps.contains(required) && c.total_bytes <= max_bytes_for_tier
            })
            .collect();

        if viable.is_empty() {
            return None;
        }
        viable.sort_by_key(|c| c.total_bytes);
        let chosen = viable[0].clone();
        let reason = format!(
            "tier={:?} ram={}MB required={:?} → {} ({} bytes)",
            device.tier,
            device.ram_bytes / (1024 * 1024),
            required.bits(),
            chosen.name,
            chosen.total_bytes
        );
        Some(ModelSelection { entry: chosen, reason })
    }
}
